packages/@cdktf/hcl2cdk/lib/coerceType.ts

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) HashiCorp, Inc
+// SPDX-License-Identifier: MPL-2.0
+import template from "@babel/template";
+import * as t from "@babel/types";
+import { logger } from "@cdktf/commons";
+import { AttributeType } from "@cdktf/provider-generator";
+import { Scope } from "./types";
+import deepEqual from "deep-equal";
+
+function changeValueAccessor(
+  ast: t.MemberExpression,
+  newAccessor: string
+): t.MemberExpression {
+  const propertyIdentifier: t.Identifier = {
+    ...(ast.property as t.Identifier),
+    name: newAccessor,
+  };
+  return {
+    ...ast,
+    property: propertyIdentifier,
+  };
+}
+
+export const coerceType = (
+  scope: Scope,
+  ast: t.Expression,
+  from: AttributeType,
+  to: AttributeType | undefined
+): t.Expression => {
+  if (to === undefined) {
+    return ast;
+  }
+
+  if (deepEqual(to, from)) {
+    return ast;
+  }
+
+  const isTerraformVariableOrLocal =
+    ast.type === "MemberExpression" &&
+    ast.property.type === "Identifier" &&
+    ast.property.name === "value" &&
+    ast.object.type === "Identifier" &&
+    Object.values(scope.variables).some(
+      (knownVars) =>
+        knownVars.variableName === (ast.object as t.Identifier).name
+    );
+
+  if (Array.isArray(to)) {
+    if (to[0] === "list" || to[0] === "set") {
+      switch (to[1]) {
+        case "string":
+          if (isTerraformVariableOrLocal) {
+            return changeValueAccessor(ast as t.MemberExpression, "listValue");
+          }
+
+          scope.hasTokenBasedTypeCoercion = true;
+          return template.expression(`cdktf.Token.asList(%%ast%%)`)({
+            ast: ast,
+          });
+        case "number":
+          scope.hasTokenBasedTypeCoercion = true;
+          return template.expression(`cdktf.Token.asNumberList(%%ast%%)`)({
+            ast: ast,
+          });
+        case "bool":
+          scope.hasTokenBasedTypeCoercion = true;
+          return template.expression(`cdktf.Token.asAny(%%ast%%)`)({
+            ast: ast,
+          });
```

Comment:
There's no `default` case for lists here – is that intended? (Map has one)

Reply:
I can't recall but I think making it explicit is best, the code is already hard enough without trying to be clever :D

Reply:
I added one now :)